### PR TITLE
Implement socket activation

### LINKED ISSUE #336: Support socket activation
It would be really useful to have Socket activation support so that speechd gets launched automatically when a client connects to the socket.

See https://github.com/brailcom/speechd/issues/335#issuecomment-625007201 for more context.

### PR DESCRIPTION
**Pull request by nteodosio**
Fixes  #336.

I followed Jhenstridge's guidelines therein, and though `spd-say` still works, I wonder if the Systemd socket activation (which is the whole point) is working; I'll be happy to follow instructions to test it in regards to that.

### CODE REVIEW

File: speech-dispatcher.service.in
Comment (alex19EP) [inline speech-dispatcher.service.in:17]: according to systemd guidelines it should be:

```suggestion
Description=Common interface to speech synthesizers
```

File: speech-dispatcher.service.in
Comment (alex19EP) [inline speech-dispatcher.service.in:27]: why is that needed?

File: speech-dispatcher.socket
Comment (alex19EP) [inline speech-dispatcher.socket:4]: maybe a good idea to set Priority=

File: src/server/speechd.c
Comment (alex19EP) [inline src/server/speechd.c:32]: and what if distribution doesn't have systemd? it should be surrounded by ifs

**Review by alex19EP — CHANGES_REQUESTED** [review]
some comments from me

**Comment by alex19EP** [comment]
> I'll be happy to follow instructions to test it in regards to that.

hello. try to disable autospaun in /etc/speech-dispatcher/speechd.conf

File: speech-dispatcher.socket
Comment (nteodosio) [reply speech-dispatcher.socket:4]: Hmm I don't know if that should have low, medium or high priority.

**Comment by nteodosio** [comment]
Thanks for the review, Alex!

> disable autospaun in /etc/speech-dispatcher/speechd.conf

I did so and it spd-say still works.

File: Makefile.am
Comment (sthibaul) [inline Makefile.am:39]: Please surround this with `if HAVE_SYSTEMD`, to fix build on systems without systemd

File: configure.ac
Comment (sthibaul) [inline configure.ac:123]: Please not make it mandatory, to fix build on systems without systemd.

Comment (sthibaul) [reply configure.ac:123]: Also, please add the libsystemd dependency to `.github/workflows/*.yml` so CI can test the code.

File: src/server/speechd.c
Comment (sthibaul) [inline src/server/speechd.c:1295]: Please mind the indentation change.

**Review by sthibaul — CHANGES_REQUESTED** [review]
This looks good overall, thanks! Just a few fixes left before I can merge it.

File: configure.ac
Comment (nteodosio) [reply configure.ac:123]: Does this look valid? Though then I assume the `AM_CONDITIONAL(HAVE_SYSTEMD, ...)` down below become unnecessary?

Comment (sthibaul) [reply configure.ac:123]: that's valid, but setting an environment variable in `configure.ac` won't make it available in `Makefile.am` and `.c` files. The canonical way is usually something like

```
PKG_CHECK_MODULES([LIBSYSTEMD], [libsystemd], [have_systemd="yes"], [have_systemd="no"])
if test x$have_systemd = xyes; then
    AC_DEFINE([HAVE_SYSTEMD], [1], [Define to 1 if you have the systemd library])
fi
AM_CONDITIONAL(HAVE_SYSTEMD, [test x$have_systemd = xyes])
```

File: .github/workflows/ci.yml
Comment (sthibaul) [inline .github/workflows/ci.yml:130]: You don't need to introduce `libsystemd0`, that'll come as a dependency from `libsystemd-dev` (and whenever the abi changes 0 would get bumped, which would require a change here, while `libsystemd-dev` alone will be fine)

**Comment by sthibaul** [comment]
It seems this is making CI fail, also

**Review by sthibaul — APPROVED** [review]

**Comment by sthibaul** [comment]
this looks good, thanks!
there is however apparently still an issue with `make distcheck`, which doesn't seem to be following the destdir of the make install check.

**Comment by nteodosio** [comment]
Thanks for your review, Samuel. I could not find much documentation about this, but in https://www.linuxfromscratch.org/blfs/view/7.4/general/bluez.html:

    --without-systemdunitdir: This switch disables installation of the systemd units.

It seems appropriate, since the error is 
```
/usr/bin/install: cannot create regular file '/usr/lib/systemd/user/speech-dispatcher.service': Permission denied
/usr/bin/install: cannot create regular file '/usr/lib/systemd/user/speech-dispatcher.socket': Permission denied
```
I must say I couldn't test this locally so far, because distcheck fails at
```
Making dvi in doc
make[2]: Entering directory '/home/q1/speechd/speech-dispatcher-0.11.2/_build/sub/doc'
  TEXI2DVI ssip.dvi
  TEXI2DVI speech-dispatcher.dvi
  TEXI2DVI spd-say.dvi
  TEXI2DVI speech-dispatcher-cs.dvi
/usr/bin/texi2dvi: etex exited with bad status, quitting.
```

**Comment by nteodosio** [comment]
Ah, I think I found it out. We ourselves even have not only systemdsystemunitdir but also systemduserunitdir.

**Comment by alex19EP** [comment]
since this pr will most likely be merged soon, I would like to raise here such a question, @sthibaul  what do you think that we package maintainers need to do in order to switch to running sd through systemd. for now I'm thinking of disabling autospawn in the system config, and enable socket for all users. Do you have any objections or suggestions?

**Comment by sthibaul** [comment]
> what do you think that we package maintainers need to do in order to switch to running sd through systemd. for now I'm thinking of disabling autospawn in the system config, and enable socket for all users. Do you have any objections or suggestions?

I don't really have an opinion. libspeechd is already able to auto-trigger the start of speechd, so this is not actually changing the behavior in that regard. AIUI it's rather for containers that the feature is useful.

**Comment by alex19EP** [comment]
> libspeechd is already able to auto-trigger the start of speechd

This is exactly what I want to prevent. Is this about autospawn config option?

**Comment by sthibaul** [comment]
> > libspeechd is already able to auto-trigger the start of speechd
> 
> This is exactly what I want to prevent. Is this about autospawn config option?

Apparently so, yes. Although that won't prevent libspeechd from starting `speech-dispatcher -a`, it will only make `speech-dispatcher -a` immediately exit.

I wonder why you want to disable auto-spawn, though?

**Comment by alex19EP** [comment]
> I wonder why you want to disable auto-spawn, though?

because then, I'll get some cool stuff:

* sd will have there own cgroup
  * I can set limits, prevent oomd from touching it
* it will be much more convenient to restart sd

**Comment by sthibaul** [comment]
> > I wonder why you want to disable auto-spawn, though?
> 
> because then, I'll get some cool stuff:
> 
>     * sd will have there own cgroup
>       
>       * I can set limits, prevent oomd from touching it
> 
>     * it will be much more convenient to restart sd

Ok, that makes sense. I was also thinking that clients can tell whether they want autospawn or not, I don't see anybody disabling it.

So I guess making a distribution enable socket activation and only that, can be a good way forward.

**Comment by lissyx** [comment]
I'm eager to test that so we can finally fix speech synthesis in Firefox Snap.

**Comment by sthibaul** [comment]
We are still getting some startup failure in the CI. That's not systematic, but that's often enough that it'll be a pain in the CI, and is a sign of some subtle bug creeping in there.

**Comment by nteodosio** [comment]
Since there is no way to retrieve the log file mentioned in the [test run](https://github.com/brailcom/speechd/actions/runs/3062916670/jobs/5067056164), I'm running it locally to gather more information, but it hangs at `1: c_api`, which doesn't look right judging by the 5 s it took to fail in the CI.

**Comment by nteodosio** [comment]
`make testinstall` hangs, but if I execute

    dbus-run-session sh -c '( pulseaudio & cd src/tests ; MALLOC_PERTURB_=1234 UBSAN_OPTIONS=halt_on_error=1 ./testsuite )'

manually, some progress: speechd starts to talk. It nonetheless failed in my first tests. I ran a make clean and tried it again, and now it says "1 completed 2 completed 3 completed ..." very fast and exits successfully...

This must be very subtle indeed.

**Comment by nteodosio** [comment]
To better understand the issue I enabled debugging on [my fork](https://github.com/nteodosio/speechd/tree/debug), but the [undefined sanitizer job](https://github.com/nteodosio/speechd/actions/runs/3129483607/jobs/5078809327) refuses to fail! :|

**Comment by sthibaul** [comment]
> Since there is no way to retrieve the log file mentioned in the [test run](https://github.com/brailcom/speechd/actions/runs/3062916670/jobs/5067056164),

It is available as artifact, on the Summary page:
https://github.com/brailcom/speechd/actions/runs/3062916670

**Comment by nteodosio** [comment]
Thanks for the new runs; Since it still very occasionally fails, I'll keep trying to get it to fail with debug activated in my fork so I can investigate why it does so.

**Comment by sthibaul** [comment]
Ok, in the end I got it failing on main as well... So no new regression after all, thus merging.

**Comment by lissyx** [comment]
\o/

**Pull request merged**
